engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb

```diff
@@ -0,0 +1,429 @@
+            ProducerCreator.new(
+              producer: producer,
+              creator: producers_user,
+              group: group
+            ).create
+          end
+          let(:producer_for_user) do
+            producer = FactoryGirl.build(:producer, name: 'Related to user')
+            ProducerCreator.new(
+              producer: producer,
+              creator: producers_user
+            ).create
+          end
+
+          before { authenticate_as producers_user }
+
+          describe 'GET #index' do
+            let!(:first_product) do
+              FactoryGirl.create(
+                :product,
+                producer: producer_for_group
+              )
+            end
+            let!(:second_product) do
+              FactoryGirl.create(
+                :product,
+                producer: producer_for_group
+              )
+            end
+
+            subject { get :index, producer_id: producer_for_group.id }
+
+            it_behaves_like 'a successful request'
+
+            describe 'its body' do
+              before { get :index, producer_id: producer_for_group.id }
+
+              subject { JSON.parse(response.body) }
+
+              its(:size) { is_expected.to eq(2) }
+              it { is_expected.to include(JSON.parse(first_product.to_json)) }
+              it { is_expected.to include(JSON.parse(second_product.to_json)) }
+            end
+          end
+
+          describe 'GET #show' do
+            let(:product) do
+              FactoryGirl.create(
+                :product,
+                producer: producer_for_group
+              )
+            end
+
+            context 'requesting a non existent product' do
+              subject { get :show, id: 666 }
+
+              it_behaves_like 'a not found request'
+            end
+
+            context 'requesting an existent product' do
+              subject { get :show, id: product.id }
+
+              it_behaves_like 'a successful request'
+
+              describe 'its body' do
+                before { get :show, id: product.id }
+
+                subject { JSON.parse(response.body) }
+
+                it { is_expected.to include(JSON.parse(product.to_json)) }
+              end
+            end
+
+            context 'requesting a product not associated to the current user' do
+              let(:another_user) { FactoryGirl.create(:user) }
+              let(:producers_user) { User.find(another_user.id) }
+
+              subject { get :show, id: product.id }
+
+              it_behaves_like 'a forbidden request'
+            end
+          end
+
+          describe 'POST #create' do
+            subject { post :create, params }
+
+            context 'when the producer is directly associated to the user' do
+              let(:params) do
+                {
+                  name: 'Tomatekken',
+                  price: 10.13,
+                  unit: Product::UNITS[:kg],
```

Comment:
IMO this :nose: a lot. In the models file already looked like a responsibility to extract, but know turns out quite evident to me. 

Here we're reaching into inner state of product and coupling to the data structure of the units. This makes it hard to change it, and by the time you realize, it won't be possible to refactor as it'll bé used everywhere.

So, we should encapsulate the data structure under a simple api